### PR TITLE
Improve init efficiency of BaseOperator for N-qubit operators

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Updates BaseOperator class to have a more efficient __init__ method for N-qubit operators. This is important for improving the performance of Cliffords and Pauli operators for large numbers of qubits which were spending a large amount of time in the initialization.

### Details and comments

Previously the initialization would construct the two input and output dimension tuples `num_qubits * (2, )`. These aren't needed if the operator is known to be an N-qubit operator,  so now these dimensions are only generated lazily for N-qubit operators when needed (such as composing them with a non-qubit operator or reshaping).

### CODE REVIEW

File: qiskit/quantum_info/operators/base_operator.py [hunk 1]
```diff
@@ -173,31 +202,57 @@ def reshape(self, input_dims=None, output_dims=None):
             QiskitError: if combined size of all subsystem input dimension or
                          subsystem output dimensions is not constant.
         """
-        clone = copy.copy(self)
+        # Trivial case
+        if num_qubits:
+            if self.num_qubits == num_qubits:
+                return self
+            dim = 2 * (2 ** num_qubits, )
+            if self.dim != dim:
```
Comment (mtreinish) [inline qiskit/quantum_info/operators/base_operator.py:210]: There might be a bit of a performance gain here for the self.num_qubits defined case here if you did `elif self.num_qubits is not None: raise` before this block but it's probably not worth the extra code complexity

**Review by mtreinish — COMMENTED** [review]
This LGTM. The only thing I think we should add here is some negative test coverage for `reshape()`. The coverage job highlighted those as a gap: https://travis-ci.com/github/Qiskit/qiskit-terra/jobs/428082573#L7275

It also would be good to see some benchmarks showing the difference with this applied. But not really worth blocking more just a curiosity thing.

File: qiskit/quantum_info/operators/base_operator.py (hunk 1 above)
Comment (chriseclectic) [reply qiskit/quantum_info/operators/base_operator.py:210]: I'm not too worried about performance in the paths that lead to an exception anyway so probably not worth extra code complexity like you say

**Review by mtreinish — APPROVED** [review]
LGTM, thanks for adding the tests

**Pull request merged**